handler/artifact_upload_handler.py

```
@@ -0,0 +1,115 @@
+"""
+A tornado application to handle uploading a conda build artifact to a specified
+file location.
+
+"""
+
+import argparse
+import hashlib
+import hmac
+import os
+import subprocess
+
+import tornado.ioloop
+import tornado.httpserver
+import tornado.web
+
+
+class ArtifactUploadHandler(tornado.web.RequestHandler):
+    """Handle writing conda build artifacts to a channel."""
+
+    def initialize(self, write_path, conda_exe, hash_expected):
+        """
+        "Hook for subclass initialization."
+
+        In this class, when we initialize `ArtifactUploadHandler`, we need to
+        set:
+         * the path that incoming files will be written to,
+         * the path to the conda executable to use for indexing the channel,
+         * the expected hashed and salted digest of the secure upload token.
+
+        """
+        self.write_path = write_path
+        self.conda_exe = conda_exe
+        self.hash_expected = hash_expected
+
+    def _check_token(self, token):
+        """
+        Check we have been passed the *correct* secure token before uploading
+        the artifact to the channel.
+        
+        """
+        salt = '42679ad04d44c96ed27470c02bfb28c3'
+        to_hash = '{}{}'.format(salt, token)
+        hash_result = hashlib.sha256(to_hash).hexdigest()
+        # Reduce the risk of timing analysis attacks.
+        return hmac.compare_digest(self.hash_expected, hash_result)
+
+    def _index_channel(self):
+        """
+        Index the conda channel at `self.write_path` to ensure the channel
+        remains consistent.
+        
+        """
+        cmds = [self.conda_exe, 'index', self.write_path]
+        subprocess.check_call(cmds)
+
+    def post(self):
+        token = self.get_argument('token')
+        file_data, = self.request.files['artifact']
+        filename = file_data['filename']
+        body = file_data['body']
+        if self._check_token(token):
```

@dkillick What if the `_check_token` fails? Appears to be silently ignored, is that right or what we want?

---

It is indeed silently ignored... because I'm not sure what this should do if it fails!

---

Hmm good question. I don't know either. 

Perhaps if we had logging enabled we'd care, right?

Could we make a relevant issue instead, querying whether we need logging, and for example, snarfing token failure might be appropriate for said logging?

You choose ...